network/p2p/ping.go

```diff
@@ -0,0 +1,195 @@
+
+const pingTimeout = time.Second * 60
+
+// PingService handles the outbound and inbound ping requests and response
+type PingService struct {
+	host           host.Host
+	pingProtocolID protocol.ID
+	version        string
+	logger         zerolog.Logger
+}
+
+func NewPingService(h host.Host, pingProtocolID protocol.ID, version string, logger zerolog.Logger) *PingService {
+	ps := &PingService{host: h, pingProtocolID: pingProtocolID, version: version, logger: logger}
+	h.SetStreamHandler(pingProtocolID, ps.PingHandler)
+	return ps
+}
+
+// PingHandler receives the inbound stream for Flow ping protocol and respond back with the PingResponse message
+func (ps *PingService) PingHandler(s network.Stream) {
+
+	errCh := make(chan error, 1)
+	defer close(errCh)
+	timer := time.NewTimer(pingTimeout)
+	defer timer.Stop()
+
+	go func() {
+		log := ps.streamLogger(s)
+		select {
+		case <-timer.C:
+			// if read or write took longer than configured timeout, then reset the stream
+			log.Error().Msg("ping timeout")
+			err := s.Reset()
+			if err != nil {
+				log.Error().Err(err).Msg("failed to reset stream")
+			}
+		case err, ok := <-errCh:
+			// reset the stream if an error occur while responding to a ping request
+			if ok {
+				log.Error().Err(err).Msg("ping response failed")
+				err := s.Reset()
+				if err != nil {
+					log.Error().Err(err).Msg("failed to reset stream")
+				}
+				return
+			}
+			// if no error, then just close the stream
+			err = s.Close()
+			if err != nil {
+				log.Error().Err(err).Msg("failed to close stream")
+			}
+		}
+	}()
+
+	pingRequest := &message.PingRequest{}
+	// create the reader
+	reader := ggio.NewDelimitedReader(s, maxPingMessageSize)
+
+	// read the request
+	err := reader.ReadMsg(pingRequest)
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	// create the writer
+	bufw := bufio.NewWriter(s)
+	writer := ggio.NewDelimitedWriter(bufw)
+
+	// create a PingResponse
+	pingResponse := &message.PingResponse{
+		Version: ps.version, // the semantic version of the build
+		// TODO populate BlockHeight:
+	}
+
+	// send the PingResponse
+	err = writer.WriteMsg(pingResponse)
+	if err != nil {
+		errCh <- err
+		return
+	}
+
+	// flush the stream
+	err = bufw.Flush()
+	if err != nil {
+		errCh <- err
+		return
+	}
+}
+
+// Ping sends a Ping request to the remote node and returns the response, rtt and error if any.
+func (ps *PingService) Ping(ctx context.Context, p peer.ID) (message.PingResponse, time.Duration, error) {
+
+	// create a done channel to indicate Ping request-response is done or an error has occurred
+	done := make(chan error, 1)
+	defer close(done)
+
+	timer := time.NewTimer(PingTimeoutSecs)
+	defer timer.Stop()
+
+	// create a new stream to the remote node
+	s, err := ps.host.NewStream(ctx, p, ps.pingProtocolID)
+	if err != nil {
+		return message.PingResponse{}, -1, fmt.Errorf("failed to create stream: %w", err)
+	}
+
+	// if ping succeeded, close the stream else reset the stream
+	go func() {
```

Comment:
Hi @yhassanzadeh13 - thanks for your question earlier about why `Reset` the stream. I got the reset statement from the libp2p's Ping logic. But after I read your question, I did some investigation and changed to code accordingly.
Libp2p's stream.Close indeed closes the stream gracefully from both ends - See this [issue](https://github.com/libp2p/go-libp2p-core/pull/166).
In a nut shell, if you would like to cause an error on the remote node while closing the stream, use `stream.Reset()` else use `stream.Close()`. 
I have modified the code to do that - if the client times out, do a reset and let the remote know that something bad has happened else just close.